classes/user/Repository.php

```diff
@@ -197,41 +196,51 @@ public function canUserAccessStage($stageId, $workflowType, $userAccessibleStage
     }
 
     /**
-     * Check for roles that give access to the passed workflow stage.
+     * Retrieve user roles which give access to (certain) submission workflow stages
+     * returns [
+     *   stage ID => [role IDs]
+     * ]
      *
-     * @param int $userId
-     * @param int $contextId
-     * @param Submission $submission
-     * @param int $stageId
-     *
-     * @return array
      */
-    public function getAccessibleStageRoles($userId, $contextId, &$submission, $stageId)
+    public function getAccessibleWorkflowStages(int $userId, int $contextId, Submission $submission, ?array $userRoleIds = null): array
     {
         $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO'); /** @var StageAssignmentDAO $stageAssignmentDao */
-        $stageAssignmentsResult = $stageAssignmentDao->getBySubmissionAndUserIdAndStageId($submission->getId(), $userId, $stageId);
+        $stageAssignmentsResult = $stageAssignmentDao->getBySubmissionAndUserIdAndStageId($submission->getId(), $userId);
+
+        if (is_null($userRoleIds)) {
+            $roleDao = DAORegistry::getDAO('RoleDAO'); /** @var RoleDAO $roleDao */
+            $userRoles = $roleDao->getByUserId($userId, $contextId);
+            $userRoleIds = [];
+            foreach ($userRoles as $userRole) { /** @var Role $userRole */
+                $userRoleIds[] = $userRole->getRoleId();
+            }
+        }
 
-        $accessibleStageRoles = [];
+        $accessibleWorkflowStages = [];
 
         // Assigned users have access based on their assignment
         while ($stageAssignment = $stageAssignmentsResult->next()) {
             $userGroup = Repo::userGroup()->get($stageAssignment->getUserGroupId());
-            $accessibleStageRoles[] = $userGroup->getRoleId();
-        }
-        $accessibleStageRoles = array_unique($accessibleStageRoles);
+            $roleId = $userGroup->getRoleId();
 
-        // If unassigned, only managers and admins have access
-        if (empty($accessibleStageRoles)) {
-            $roleDao = DAORegistry::getDAO('RoleDAO'); /** @var RoleDAO $roleDao */
-            if ($roleDao->userHasRole($contextId, $userId, Role::ROLE_ID_MANAGER)) {
-                $accessibleStageRoles[] = Role::ROLE_ID_MANAGER;
+            // Check global user roles within the context, e.g., user can be assigned in the role, which was revoked
+            if (!in_array($roleId, $userRoleIds)) {
+                continue;
             }
-            if ($roleDao->userHasRole(Application::CONTEXT_SITE, $userId, Role::ROLE_ID_SITE_ADMIN)) {
-                $accessibleStageRoles[] = Role::ROLE_ID_SITE_ADMIN;
+
+            $accessibleWorkflowStages[$stageAssignment->getStageId()][] = $roleId;
+        }
+
+        // Managers and admin have access if not assigned to the submission or are assigned in a revoked role
+        $managerRoles = array_intersect($userRoleIds, [Role::ROLE_ID_SITE_ADMIN, Role::ROLE_ID_MANAGER]);
```

Comment:
I think `$userRoleIds` will never contain `Role::ROLE_ID_SITE_ADMIN` when the roles are fetched within the function, since that is a site-level role (`context_id = 0`) and `$userRoleIds` is populated from a specific `$contextId`.

Reply:
Thanks, fixed! It now checks for the admin role in a proper way.